Mathlib/Computability/AkraBazzi/AkraBazzi.lean

```diff
@@ -0,0 +1,1485 @@
+    T =O[atTop] (fun n => (1 - ε n) * asympBound g a b n) := by
+  let b' := b R.min_bi / 2
+  have hb_pos : 0 < b' := R.bi_min_div_two_pos
+  rw [isBigO_atTop_iff_eventually_exists]
+  obtain ⟨c₁, hc₁, h_sumTransform_aux⟩ := R.eventually_atTop_sumTransform_ge
+  filter_upwards [eventually_ge_atTop R.n₀,       -- n₀_ge_Rn₀
+      eventually_forall_ge_atTop.mpr eventually_one_sub_smoothingFn_pos,    -- h_smoothing_pos
+      eventually_forall_ge_atTop.mpr
+        <| eventually_one_sub_smoothingFn_gt_const (1/2) (by norm_num),    -- h_smoothing_gt_half
+      eventually_forall_ge_atTop.mpr R.eventually_asympBound_pos,            -- h_asympBound_pos
+      eventually_forall_ge_atTop.mpr R.eventually_asympBound_r_pos,          -- h_asympBound_r_pos
+      (tendsto_nat_floor_mul_atTop b' hb_pos).eventually_forall_ge_atTop
+        R.eventually_asympBound_pos,   -- h_asympBound_floor
+      eventually_gt_atTop 0,                                                -- n₀_pos
+      eventually_forall_ge_atTop.mpr R.eventually_one_sub_smoothingFn_r_pos,  -- h_smoothing_r_pos
+      eventually_forall_ge_atTop.mpr R.rpow_p_mul_one_sub_smoothingFn_le,    -- bound1
+      (tendsto_nat_floor_mul_atTop b' hb_pos).eventually_forall_ge_atTop
+        eventually_one_sub_smoothingFn_pos,   -- h_smoothingFn_floor
+      eventually_forall_ge_atTop.mpr h_sumTransform_aux,                     -- h_sumTransform
+      eventually_forall_ge_atTop.mpr R.eventually_bi_mul_le_r]               -- h_bi_le_r
+    with n₀ n₀_ge_Rn₀ h_smoothing_pos h_smoothing_gt_half
+      h_asympBound_pos h_asympBound_r_pos h_asympBound_floor n₀_pos h_smoothing_r_pos
+      bound1 h_smoothingFn_floor h_sumTransform h_bi_le_r
+  -- Max of the ratio `T(n) / asympBound(n)` over the base case `n ∈ [b * n₀, n₀)`
+  have h_base_nonempty := R.base_nonempty n₀_pos
+  let base_max : ℝ :=
+    (Finset.Ico (⌊b' * n₀⌋₊) n₀).sup' h_base_nonempty
+      fun n => T n / ((1 - ε n) * asympBound g a b n)
+  -- The big-O constant we are aiming for: max of the base case ratio and what we need to
+  -- cancel out the `g(n)` term in the calculation below
+  set C := max (2 * c₁⁻¹) base_max with hC
+  refine ⟨C, fun n hn => ?_⟩
+  -- Base case: statement is true for `b' * n₀ ≤ n < n₀`
+  have h_base : ∀ n ∈ Finset.Ico (⌊b' * n₀⌋₊) n₀, T n ≤ C * ((1 - ε n) * asympBound g a b n) := by
+    intro n hn
+    rw [Finset.mem_Ico] at hn
+    have htmp1 : 0 < 1 - ε n := h_smoothingFn_floor n hn.1
+    have htmp2 : 0 < asympBound g a b n := h_asympBound_floor n hn.1
+    rw [←_root_.div_le_iff (by positivity)]
+    rw [←Finset.mem_Ico] at hn
+    calc T n / ((1 - ε ↑n) * asympBound g a b n)
+           ≤ (Finset.Ico (⌊b' * n₀⌋₊) n₀).sup' h_base_nonempty
+                (fun z => T z / ((1 - ε z) * asympBound g a b z)) :=
+                  Finset.le_sup'_of_le _ (b := n) hn <| le_refl _
+         _ ≤ C := le_max_right _ _
+  have h_asympBound_pos' : 0 < asympBound g a b n := h_asympBound_pos n hn
```

Comment:
You have a comment about the base case above. Could you put another comment around here saying that you will now conclude all cases by induction?
Same thing in next proof.